scipy/spatial/_lloyd_algorithm.py

```diff
@@ -0,0 +1,234 @@
+
+        # move points towards centroids:
+        # Centroid in n-D is the mean for uniformly distributed nodes
+        # of a geometry.
+        centroid = np.mean(verts, axis=0)
+        new_points[ii] = points[ii] + (centroid - points[ii]) * decay
+
+    # only update points to centroid within the region
+    is_valid = np.all(np.logical_and(new_points >= 0, new_points <= 1), axis=1)
+    points[is_valid] = new_points[is_valid]
+
+    return points
+
+
+def lloyd_centroidal_voronoi_tessellation(
+        points: npt.ArrayLike,
+        *,
+        tol: DecimalNumber = 1e-5,
+        maxiter: IntNumber = 10,
+        decay: Optional[npt.ArrayLike] = None,
+        qhull_options: Optional[str] = None,
+) -> np.ndarray:
+    """Approximate Centroidal Voronoi Tessellation.
+
+    Perturb points in :math:`[0, 1]^d` using Lloyd-Max algorithm.
+
+    Parameters
+    ----------
+    points : array_like (n, d)
+        The points to iterate on. With ``n`` the number of points and ``d``
+        the dimension.
+    tol : float, optional
+        Tolerance for termination. If the min of the L1-norm over the points
+        changes less than `tol`, it stops the algorithm. Default is 1e-5.
+    maxiter : int, optional
+        Maximum number of iterations. It will stop the algorithm even if
+        `tol` is above the threshold.
+        Too many iterations tend to cluster the points as a hypersphere.
+        Default is 10.
+    decay : {float, array_like (maxiter)}, optional
+        Relaxation decay. A positive value would move the points toward
+        their centroid, and negative value would move them away.
+        1 would move the points to their centroid. Default is a varying
+        exponential decay starting at 2 and ending at 1 after `maxiter`.
+    qhull_options : str, optional
+        Additional options to pass to Qhull. See Qhull manual
+        for details. (Default: "Qbb Qc Qz Qj Qx" for ndim > 4 and
+        "Qbb Qc Qz Qj" otherwise.)
+
+    Returns
+    -------
+    points : array_like (n, d)
+        The points after being processed by Lloyd-Max algorithm.
+
+    Notes
+    -----
+    Lloyd-Max algorithm is an iterative process with the purpose of improving
+    the dispersion of points. For given points: (i) compute a Voronoi
+    Tessellation; (ii) find the centroid of each Voronoi cell; (iii) move the
+    points toward the centroid of their respective cell. See [1]_, [2]_.
+
+    The process converges to equally spaced points. It implies that measures
+    like the discrepancy could suffer from too many iterations. On the other
+    hand, L1 and L2 distances should improve. This is especially true with
+    QMC methods which tend to favor the discrepancy over other criteria.
+
+    .. note::
+
+        The current implementation does not intersect the Voronoi Tessellation
+        with the boundaries. This implies that for a low number of points,
+        empirically below 20, no Voronoi cell is touching the boundaries.
```

Comment:
This isn't really empirically accurate though--there are an infinite number of cases at arbitrarily high number of points where the current algorithm will fail. Consider any of the convex n-polygons with n between `3` and approaching infinity. For example, generate `2000` random points on the circumference of a circle to produce a 2000-vertex convex polygon that will collapse inward on the latest version of this branch:

![test_convex_polygon](https://user-images.githubusercontent.com/7903078/144038741-3fa8ea12-1e69-451a-a5d4-af780d8f0734.png)

I still think this PR/algorithm is very much in an empirical testing phase rather than a production-ready phase, so I think we should delay to the next milestone unless the testing/implementation/limitation descriptions are more robust.

If you really want to be robust in describing the limitations, you may for example need to assess the percentage of input generators that can lie on the convex hull before the algorithm starts to fail. That seems like a tricky thing to assess robustly.

Reply:
(and I still think that random point injection near boundaries has merit over failing on basically all convex inputs if you aren't going to implement edge intersection at infinity)

Reply:
Shouldn't the `decay` parameter be able to prevent the total collapse of the circle example?

Reply:
Pretty diagram. I am sure we can always find corner cases like these. I still think that, provided we word correctly the limitations, this would still be useful as it is.

Reply:
It isn't really a corner case, it is every convex polyhedron in the universe I think?

Perhaps more useful might be the robust handling for 2-D and 3-D with intersection, then at a later time try to expand to N-D.

Reply:
I think we don't intent to use this function the same way. What would you use it for?

Since we don't handle boundaries, to me it's only helpful to spread points in a hypercube. If that's the goal, then I would expect users to provide something which is already more or less covering the space.

Reply:
I'm mostly saying it *should* be able to spread points in general, and for lower dimensions the literature suggests this is possible with edge intersection.

An implementation that explicitly does the opposite for an almost infinite set of possible inputs doesn't seem right to me, even if it isn't the input that you personally intended the function to tolerate.

Reply:
I am just afraid that this will never happen... I am not sure I would have time nor be able to come up with something. And IMHO my current limitations should not block anything. Some future user might be able to improve this. But if nothing is there, the entry barrier is higher. You need to discover first that you can do this, then use it, fail, start to code something maybe, then realise there was this PR, etc.

Reply:
I'm partial to tupui's interpretation of the function as is. I'm not sure why you'd want to use Lloyd's algorithm on a convex polyhedron? 

I haven't looked closely at the `decay` logic but I think it should be the bulwark against degenerate cases.

Reply:
My point is that in the real world data is not necessarily random and could be structured. A convex structure in n-dimensions is certainly something that could pop out of some kind of workflow that one might have.

I'd also point out that the [wikipedia article](https://en.wikipedia.org/wiki/Lloyd%27s_algorithm) uses a convex polygon for its example of iteration. If it really is that unusual, I doubt it would be used in an example.

![image](https://user-images.githubusercontent.com/7903078/144165081-b1ea0a6d-41da-4af0-9b3b-f9c18bb1f3a2.png)
![image](https://user-images.githubusercontent.com/7903078/144165117-e9028558-f185-4196-9bc5-f449150038d8.png)
![image](https://user-images.githubusercontent.com/7903078/144165141-68613ce3-bc55-4ed2-b9a3-f588d23f4428.png)
![image](https://user-images.githubusercontent.com/7903078/144165156-238044fb-c679-4142-8a4c-74c7dd8b647d.png)

I also thought that the description in the notes above was pretty misleading--you can't really count on the number of points to be a sane indicator of the performance of the algorithm if you don't even discuss the shape of the input. Some careful thought there re: convexity considerations and so on might have swayed me, but I dunno, I still think it should be done a bit more robustly.

I also think it wouldn't be that hard to simply use `ConvexHull` to check if all points in the input are members of the convex hull, and then raise a helpful error message (`the algorithm is currently incapable of separating an input set on the convex hull`). But I didn't really see the kind of detailed analysis/testing in this regard that I was looking for here, and I think that silently doing the wrong thing doesn't make sense if we know we can't handle specific cases.

Reply:
It is also noteworthy that a few hours of analysis might reveal that the only really substantial problems are convex hull cases, but I haven't really seen this analysis done. You could write tests for this easily enough--if a bunch of random hull cases in N-D become tractable for separation by addition of only a single internal point you're in pretty good shape, if not it is messy.